### PR TITLE
Extract udp crate

### PR DESCRIPTION
**Pull request by kpp**
This one is easy - moved code from quinn/platform into quinn-udp subcrate.
I will try to apply the code from https://github.com/ipfs-rust/udp-socket (which is runtime-agnostic due to use of `async-io` instead of `tokio`) in the [next PR](https://github.com/quinn-rs/quinn/pull/1183).

Fix https://github.com/quinn-rs/quinn/issues/1178

### CODE REVIEW

File: quinn-udp/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "quinn-udp"
+version = "0.7.0"
+authors = ["Benjamin Saunders <ben.e.saunders@gmail.com>", "Dirkjan Ochtman <dirkjan@ochtman.nl>"]
+license = "MIT OR Apache-2.0"
+repository = "https://github.com/djc/quinn"
```
Comment (djc) [inline quinn-udp/Cargo.toml:6]: Should use the quinn-rs URL here.

File: quinn-udp/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+all-features = true
+
+[badges]
+maintenance = { status = "experimental" }
+
+[dependencies]
+futures-util = { version = "0.3.11", features = ["io"] }
+libc = "0.2.69"
+mio = { version = "0.7.7", features = ["net"] }
+proto = { package = "quinn-proto", path = "../quinn-proto", version = "0.7" }
```
Comment (djc) [inline quinn-udp/Cargo.toml:23]: Feels weird for this crate to depend on proto. Why is that? (Hard to see since there are no diffs for the moved files.)

**Review by djc — COMMENTED** [review]
Looks good, some comments.

File: quinn-udp/Cargo.toml (hunk 2 above)
Comment (kpp) [reply quinn-udp/Cargo.toml:23]: That's because of:

> use proto::{EcnCodepoint, Transmit};

File: quinn-udp/Cargo.toml (hunk 1 above)
Comment (kpp) [reply quinn-udp/Cargo.toml]: Done. Also fixed repo url in quinn and quinn-proto.

File: quinn-udp/Cargo.toml (hunk 2 above)
Comment (Ralith) [reply quinn-udp/Cargo.toml:23]: Per discussion in #1178 we're expressly intending for this crate to be used only with quinn-proto, so this seems reasonable to me.

**Review by Ralith — DISMISSED** [review]
Thanks, this generally LGTM. Sorry for the long review turnaround!

I appreciate the logically isolated commits; they make it a lot easier to scan things. Would you mind squashing the first two together so that there's not an unbuildable state between them?

**Comment by kpp** [comment]
Done. I also rebased onto main.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks!

**Pull request merged**